File: Codes_morgan/TOURE/Modelisation_numerique/Codes Volumes Finis - Eulérien et Lagrangien-20260129/Python/Lagrange_1D.py
import numpy as np
import matplotlib.pyplot as plt

def Lagrange_1D(pb, flx_type, nc, cfl, vplt):
    if vplt == 1:
        plt.close('all')

    if pb == 1:
        xd, xf, xdis = 0, 1, 0.5
        gam, tfin = 1.4, 0.2
        den_L, pre_L, vit_L = 1, 1, 0
        den_R, pre_R, vit_R = 0.125, 0.1, 0
    elif pb == 2:
        xd, xf, xdis = 0, 1, 0.5
        gam, tfin = 1.4, 0.15
        den_L, pre_L, vit_L = 1, 0.4, -2
        den_R, pre_R, vit_R = 1, 0.4, 2

    dx = (xf - xd) / nc
    xint = np.linspace(xd, xf, nc + 1)
    xi = np.arange(xd + dx / 2, xf - dx / 2 + dx / 2, dx)

    den = np.zeros(nc)
    vit = np.zeros(nc)
    pre = np.zeros(nc)
    ene = np.zeros(nc)
    cel = np.zeros(nc)
    masse = np.zeros(nc)
    Vsol = np.zeros((nc, 3))
    
    for ii in range(nc):
        if xint[ii + 1] <= xdis:
            den[ii], pre[ii], vit[ii] = den_L, pre_L, vit_L
        else:
            den[ii], pre[ii], vit[ii] = den_R, pre_R, vit_R
        
        ene[ii] = pre[ii] / (den[ii] * (gam - 1)) + vit[ii] ** 2 / 2
        cel[ii] = np.sqrt(pre[ii] * gam / den[ii])
        masse[ii] = den[ii] * dx

        Vsol[ii, 0] = 1 / den[ii]
        Vsol[ii, 1] = vit[ii]
        Vsol[ii, 2] = ene[ii]

    tps = 0
    flux_int = np.zeros((nc + 1, 3))
    n_it = 0

    while tps < tfin:
        dt = cfl * np.min(masse) / np.max(den * cel)
        if tps + dt > tfin + 1e-6:
            dt = tfin - tps
        
        flux_int[0, :] = flux_num(Vsol[0, :], Vsol[0, :], gam, flx_type)
        xint[0] -= dt * flux_int[0, 0]
        
        for ii in range(nc):
            if ii < nc - 1:
                flux_int[ii + 1, :] = flux_num(Vsol[ii, :], Vsol[ii + 1, :], gam, flx_type)
            else:
                flux_int[ii + 1, :] = flux_num(Vsol[ii, :], Vsol[ii, :], gam, flx_type)
            
            Vsol[ii, :] -= dt / masse[ii] * (flux_int[ii + 1, :] - flux_int[ii, :])
            xint[ii + 1] -= dt * flux_int[ii + 1, 0]
            xi[ii] = (xint[ii] + xint[ii + 1]) / 2
            
            den[ii] = 1 / Vsol[ii, 0]
            vit[ii] = Vsol[ii, 1]
            pre[ii] = den[ii] * (gam - 1) * (Vsol[ii, 2] - 0.5 * vit[ii] ** 2)
            cel[ii] = np.sqrt(pre[ii] * gam / den[ii])

        tps += dt
        n_it += 1

    if pb == 1:
        file_den_sol = 'Sol_exactes/sod_den.dat'
        file_eni_sol = 'Sol_exactes/sod_eni.dat'
    elif pb == 2:
        file_den_sol = 'Sol_exactes/dbl_det_den.dat'
        file_eni_sol = 'Sol_exactes/dbl_det_eni.dat'

    with open(file_den_sol, 'r') as fid_den, open(file_eni_sol, 'r') as fid_eni:
        TT = fid_den.read()
        ss = sum(c == '\n' for c in TT)
        fid_den.seek(0)
        
        xs = np.zeros(ss)
        den_sol = np.zeros(ss)
        eni_sol = np.zeros(ss)
        
        for ii in range(ss):
            a_den = np.array([float(val) for val in fid_den.readline().split()])
            a_eni = np.array([float(val) for val in fid_eni.readline().split()])
            
            xs[ii] = a_den[0]
            den_sol[ii] = a_den[1]
            eni_sol[ii] = a_eni[1]

    if vplt == 1:
        plt.figure(1, figsize=(10, 8))
        plt.plot(xs, den_sol, '-r', linewidth=3)
        plt.plot(xi, den, '-db')
        plt.xlabel('x')
        plt.ylabel('densité')
        plt.title("Densité à l'instant final")
        plt.legend(['solution exacte', 'schéma Lagrangien'], loc='upper right')

        plt.figure(2, figsize=(10, 8))
        plt.plot(xs, eni_sol, '-r', linewidth=3)
        plt.plot(xi, pre / ((gam - 1) * den), '-db')
        plt.xlabel('x')
        plt.ylabel('énergie interne')
        plt.title("Énergie interne à l'instant final")
        plt.legend(['solution exacte', 'schéma Lagrangien'], loc='upper left')
    else:
        return xs,den_sol,eni_sol,xi,den,pre / ((gam - 1) * den)

    plt.show()


def flux_num(VL, VR, gam, flx_type):
    den_L = 1 / VL[0]
    vit_L = VL[1]
    ene_L = VL[2]
    pre_L = den_L * (gam - 1) * (VL[2] - 0.5 * VL[1] ** 2)
    cel_L = np.sqrt(pre_L * gam / den_L)

    den_R = 1 / VR[0]
    vit_R = VR[1]
    ene_R = VR[2]
    pre_R = den_R * (gam - 1) * (VR[2] - 0.5 * VR[1] ** 2)
    cel_R = np.sqrt(pre_R * gam / den_R)

    if flx_type == 1:
        Z0 = max(den_L * cel_L, den_R * cel_R)
        vit_riemann = 0.5 * (vit_L + vit_R + Z0 * (1 / den_R - 1 / den_L))
        pre_riemann = 0.5 * (pre_L + pre_R - Z0 * (vit_R - vit_L))
        pre_vit_riemann = 0.5 * (pre_L * vit_L + pre_R * vit_R - Z0 * (ene_R - ene_L))
    elif flx_type == 2:
        Z0 = max(den_L * cel_L, den_R * cel_R)
        vit_riemann = 0.5 * (vit_L + vit_R - (pre_R - pre_L) / Z0)
        pre_riemann = 0.5 * (pre_L + pre_R - Z0 * (vit_R - vit_L))
        pre_vit_riemann = pre_riemann * vit_riemann
    else:
        ZL = den_L * cel_L
        ZR = den_R * cel_R
        vit_riemann = (ZL * vit_L + ZR * vit_R - (pre_R - pre_L)) / (ZL + ZR)
        pre_riemann = (ZR * pre_L + ZL * pre_R - ZL * ZR * (vit_R - vit_L)) / (ZL + ZR)
        pre_vit_riemann = pre_riemann * vit_riemann
    
    return [-vit_riemann, pre_riemann, pre_vit_riemann]


## Voici un exemple de comment appeler la funtion
## Lagrange_1D(1, 3, 100, 0.9, 1)

File: Codes_morgan/TOURE/Modelisation_numerique/Codes Volumes Finis - Eulérien et Lagrangien-20260129/Matlab/Lagrange_1D.m
function Lagrange_1D(pb,flx_type,nc,cfl,plt)

if (plt==1)
    close all;
end

if (pb==1)
    xd=0;
    xf=1;
    xdis=0.5;
    gam=1.4;
    tfin=0.2;
    den_L=1;
    pre_L=1;
    vit_L=0;
    den_R=0.125;
    pre_R=0.1;
    vit_R=0;
elseif (pb==2)
    xd=0;
    xf=1;
    xdis=0.5;
    gam=1.4;
    tfin=0.15;
    den_L=1;
    pre_L=0.4;
    vit_L=-2;
    den_R=1;
    pre_R=0.4;
    vit_R=2;
end

dx=(xf-xd)/nc;
xint=linspace(xd,xf,nc+1);
xi=xd+dx/2:dx:xf-dx/2;

den=zeros(nc,1);
vit=zeros(nc,1);
pre=zeros(nc,1);
ene=zeros(nc,1);
cel=zeros(nc,1);
masse=zeros(nc,1);
Vsol=zeros(nc,3);
for ii=1:nc    
    if (xint(ii+1)<=xdis)
        den(ii)=den_L;
        pre(ii)=pre_L;
        vit(ii)=vit_L;
    else
        den(ii)=den_R;
        pre(ii)=pre_R;
        vit(ii)=vit_R;
    end
    
    ene(ii)=pre(ii)/(den(ii)*(gam-1))+vit(ii)^2/2;
    cel(ii)=sqrt(pre(ii)*gam/den(ii));
    masse(ii)=den(ii)*dx;

    Vsol(ii,1)=1/den(ii);
    Vsol(ii,2)=vit(ii);
    Vsol(ii,3)=ene(ii);
end

tps=0;

flux_int=zeros(nc+1,3);

n_it=0;
while (tps<tfin)   

    dt=cfl*min(masse)/max(den.*cel);
    if (tps+dt>tfin+1.D-6)
        dt=tfin-tps;
    end
    
    flux_int(1,:)=flux_num(Vsol(1,:),Vsol(1,:),gam,flx_type);
    xint(1)=xint(1)-dt*flux_int(1,1);
    for ii=1:nc
        
        if (ii<nc)
            flux_int(ii+1,:)=flux_num(Vsol(ii,:),Vsol(ii+1,:),gam,flx_type);
        else
            flux_int(ii+1,:)=flux_num(Vsol(ii,:),Vsol(ii,:),gam,flx_type);
        end
        
        Vsol(ii,:)=Vsol(ii,:)-dt/masse(ii)*(flux_int(ii+1,:)-flux_int(ii,:));
        xint(ii+1)=xint(ii+1)-dt*flux_int(ii+1,1);
        xi(ii)=(xint(ii)+xint(ii+1))/2;
        
        den(ii)=1/Vsol(ii,1);
        vit(ii)=Vsol(ii,2);
        pre(ii)=den(ii)*(gam-1)*(Vsol(ii,3)-0.5*vit(ii)^2);
        cel(ii)=sqrt(pre(ii)*gam/den(ii));
    end
    
    tps=tps+dt;
    n_it=n_it+1;
    
end

if (pb==1)
    file_den_sol='Sol_exactes/sod_den.dat';
    file_eni_sol='Sol_exactes/sod_eni.dat';
elseif (pb==2)
    file_den_sol='Sol_exactes/dbl_det_den.dat';
    file_eni_sol='Sol_exactes/dbl_det_eni.dat';
end
fid_den = fopen(file_den_sol,'r');
TT = fread(fid_den);
ss=sum(TT==10);
fclose(fid_den);

fid_den = fopen(file_den_sol,'r');
fid_eni = fopen(file_eni_sol,'r');
xs=zeros(1,ss);
den_sol=zeros(1,ss);
eni_sol=zeros(1,ss);
for ii=1:ss   
  a_den=fscanf(fid_den,'%g %g',[2 1]);
  a_eni=fscanf(fid_eni,'%g %g',[2 1]);
  xs(ii)=a_den(1);
  den_sol(ii)=a_den(2);
  eni_sol(ii)=a_eni(2);
end

if (plt==1)
    fig1=figure(1);
    rect=[100 -100 800 600];
    set(fig1,'Position',rect);
    plot(xs,den_sol,'-r','linewidth',2);
    hold on
    plot(xi,den,'-dk');

    xlabel('x');
    ylabel('densité');
    title("Densité à l'instant final");
    legend('solution exacte','schéma Lagrangien','Location','northeast');

    fig2=figure(2);
    rect=[1000 -100 800 600];
    set(fig2,'Position',rect);
    plot(xs,eni_sol,'-r','linewidth',2);
    hold on
    plot(xi,pre./((gam-1)*den),'-dk');

    xlabel('x');
    ylabel('énergie interne');
    title("Énergie interne à l'instant final");
    legend('solution exacte','schéma Lagrangien','Location','northwest');
else
    fig1=figure(1);
    plot(xi,den,'-dk');

    fig2=figure(2);
    plot(xi,pre./((gam-1)*den),'-dk');
end

end


function FF=flux_num(VL,VR,gam,flx_type)
        
den_L=1/VL(1);
vit_L=VL(2);
ene_L=VL(3);
pre_L=den_L*(gam-1)*(VL(3)-0.5*VL(2)^2);
cel_L=sqrt(pre_L*gam/den_L);

den_R=1/VR(1);
vit_R=VR(2);
ene_R=VR(3);
pre_R=den_R*(gam-1)*(VR(3)-0.5*VR(2)^2);
cel_R=sqrt(pre_R*gam/den_R);


if (flx_type==1)
    Z0=max(den_L*cel_L,den_R*cel_R);
    
    vit_riemann=0.5*(vit_L+vit_R+Z0*(1/den_R-1/den_L));
    pre_riemann=0.5*(pre_L+pre_R-Z0*(vit_R-vit_L));
    pre_vit_riemann=0.5*(pre_L*vit_L+pre_R*vit_R-Z0*(ene_R-ene_L));
elseif (flx_type==2)
    Z0=max(den_L*cel_L,den_R*cel_R);
    
    vit_riemann=0.5*(vit_L+vit_R-(pre_R-pre_L)/Z0);
    pre_riemann=0.5*(pre_L+pre_R-Z0*(vit_R-vit_L));
    pre_vit_riemann=pre_riemann*vit_riemann;
else
    ZL=den_L*cel_L;
    ZR=den_R*cel_R;
    
    vit_riemann=(ZL*vit_L+ZR*vit_R-(pre_R-pre_L))/(ZL+ZR);
    pre_riemann=(ZR*pre_L+ZL*pre_R-ZL*ZR*(vit_R-vit_L))/(ZL+ZR);
    pre_vit_riemann=pre_riemann*vit_riemann;
end

FF(1)=-vit_riemann;
FF(2)=pre_riemann;
FF(3)=pre_vit_riemann;

end